### PR TITLE
Field mapping improvements

### PR DESCRIPTION
**Pull request by greg0ire**
It will be hard to finish https://github.com/doctrine/orm/pull/10405, but this should get us closer to it.
Also, I learnt that DTOs shouldn't be used in mapping drivers, because drivers produce incomplete field mappings (they might not have a type), and also produce mapping overrides that might have even fewer fields. DTOs should be used in the class metadata and its consumers.

### CODE REVIEW

**Comment by greg0ire** [comment]
@michnovka please review

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 1]
```diff
@@ -1584,7 +1586,7 @@ private function isTypedProperty(string $name): bool
      *
      * @param  array{fieldName: string, type?: mixed} $mapping The field mapping to validate & complete.
```
Comment (michnovka) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: if we update return type, we should update the `type?: string` here in param too

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 2]
```diff
@@ -1628,7 +1630,7 @@ private function validateAndCompleteTypedAssociationMapping(array $mapping): arr
      *     enumType?: class-string,
      * } $mapping The field mapping to validate & complete.
      *
-     * @return mixed[] The updated mapping.
+     * @return FieldMapping The updated mapping.
```
Comment (michnovka) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:1633]: FieldMapping type has a compulsory array-key `columnName`. I am not sure that this is satisfied here, as the function does not set it and @param $mapping does not specify it.

**Review by michnovka — CHANGES_REQUESTED** [review]
have a look at my review

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:1633]: It's satisfied right below the comment "Complete fieldName and columnName mapping"

Comment (michnovka) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:1633]: Thats a different function tho `validateAndCompleteFieldMapping`. The change you made is for `validateAndCompleteTypedAssociationMapping` and that one does NOT return `columnName` in the array

Comment (michnovka) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:1633]: sorry, my bad I was working on some badly lined version in my local repo.

**Review by michnovka — APPROVED** [review]
looks good now

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:1633]: No worries, and thanks for reviewing 👍

**Review by SenseException — APPROVED** [review]

**Pull request merged**